rl4co/envs/routing/pdp/env.py

```diff
@@ -209,3 +209,317 @@ def check_solution_validity(self, td, actions):
+            ),
+            rec_current=UnboundedDiscreteTensorSpec(
+                shape=(self.generator.num_loc + 1),
+                dtype=torch.int64,
+            ),
+            rec_best=UnboundedDiscreteTensorSpec(
+                shape=(self.generator.num_loc + 1),
+                dtype=torch.int64,
+            ),
+            visited_time=UnboundedDiscreteTensorSpec(
+                shape=(self.generator.num_loc + 1, self.generator.num_loc + 1),
+                dtype=torch.int64,
+            ),
+            action_record=UnboundedDiscreteTensorSpec(
+                shape=(self.generator.num_loc + 1, self.generator.num_loc + 1),
+                dtype=torch.int64,
+            ),
+            i=UnboundedDiscreteTensorSpec(
+                shape=(1),
+                dtype=torch.int64,
+            ),
+            action_mask=UnboundedDiscreteTensorSpec(
+                shape=(self.generator.num_loc + 1, self.generator.num_loc + 1),
+                dtype=torch.bool,
+            ),
+            shape=(),
+        )
+        self.action_spec = BoundedTensorSpec(
+            shape=(3,),
+            dtype=torch.int64,
+            low=0,
+            high=self.generator.num_loc + 1,
+        )
+        self.reward_spec = UnboundedContinuousTensorSpec(shape=(1,))
+        self.done_spec = UnboundedDiscreteTensorSpec(shape=(1,), dtype=torch.bool)
+
+    @staticmethod
+    def _get_reward(td, actions) -> TensorDict:
+        # This function is not used for improvement tasks since the reward is computed per step
+        assert False
+
+    def check_solution_validity(self, td, actions):
+        pass
+        #Todo
+        
+    @staticmethod
+    def get_mask(selected_node, td):
+        
+        visited_time = td["visited_time"]
+        bs, gs = visited_time.size()
+        visited_time = visited_time % gs
+        arange = torch.arange(bs)
+        
+        visited_order_map = visited_time.view(bs, gs, 1) > visited_time.view(bs, 1, gs)
+        mask = visited_order_map.clone()
+        mask[arange, selected_node.view(-1)] = True
+        mask[arange, selected_node.view(-1) + gs // 2] = True
+        mask[arange,:,selected_node.view(-1)] = True
+        mask[arange,:,selected_node.view(-1) + gs // 2] = True
+        
+        bs, gs, _ = visited_order_map.size()
+        
+        return mask
+
+    @staticmethod
+    def get_costs(coordinates, rec):
+        
+        batch_size, size = rec.size()
+        
+        # calculate the route length value
+        d1 = coordinates.gather(1, rec.long().unsqueeze(-1).expand(batch_size, size, 2))
+        d2 = coordinates
+        length =  (d1  - d2).norm(p=2, dim=2).sum(1)
+        
+        return length
+
+    @staticmethod
+    def _random_action(td):
+        batch_size, graph_size = td["rec_best"].size()
+        selected_node = ((torch.rand(batch_size,1) * graph_size//2) % (graph_size//2)).long()
+        mask = PDPRuinRepairEnv.get_mask(selected_node + 1, td)
```

Comment:
No, the random_policy is functional based on the td["action_mask"] of constructive MDP. For the improvement case, the masking and getting random action are more complicated. But this function is used for code tests only and I think it will not be used by the training or inference (or we can remove this function?).

Reply:
I think it's fine then for now, np